File: check_d_lamp_dom.js
const fs = require('fs');
const path = require('path');

try {
    const jsPath = path.join(__dirname, 'd-lamp.js');
    const htmlPath = path.join(__dirname, 'd-lamp.html');
    
    const jsContent = fs.readFileSync(jsPath, 'utf8');
    const htmlContent = fs.readFileSync(htmlPath, 'utf8');

    // Extract all document.getElementById("...")
    const jsIds = new Set();
    const idRegex = /document\.getElementById\(['"]([a-zA-Z0-9_-]+)['"]\)/g;
    let match;
    while ((match = idRegex.exec(jsContent)) !== null) {
        jsIds.add(match[1]);
    }

    // Extract all id="..." in HTML
    const htmlIds = new Set();
    const htmlIdRegex = /id=['"]([a-zA-Z0-9_-]+)['"]/g;
    while ((match = htmlIdRegex.exec(htmlContent)) !== null) {
        htmlIds.add(match[1]);
    }

    const missingInHtml = [...jsIds].filter(id => !htmlIds.has(id));

    console.log(`========================================`);
    console.log(`DOM INTEGRITY REPORT: d-lamp Mode`);
    console.log(`========================================`);
    console.log(`Total Unique IDs queried in JS: ${jsIds.size}`);
    console.log(`Total Unique IDs defined in HTML: ${htmlIds.size}`);
    console.log(`----------------------------------------`);

    if (missingInHtml.length > 0) {
        console.error(`❌ CRITICAL FAILURE: JS references these IDs, but they do NOT exist in HTML:`);
        missingInHtml.forEach(id => console.error(`  - ${id}`));
        process.exit(1);
    } else {
        console.log(`✅ SUCCESS: All JS DOM element query IDs are present in the HTML DOM!`);
        process.exit(0);
    }
} catch (error) {
    console.error(`❌ FATAL ERROR running verification:`, error.message);
    process.exit(1);
}

File: check_d_lamp_dom.py
import re
import os
import sys

try:
    current_dir = os.path.dirname(os.path.abspath(__file__))
    js_path = os.path.join(current_dir, 'd-lamp.js')
    html_path = os.path.join(current_dir, 'd-lamp.html')
    
    with open(js_path, "r", encoding="utf-8") as f:
        js_content = f.read()

    with open(html_path, "r", encoding="utf-8") as f:
        html_content = f.read()

    # Extract all document.getElementById("...") or '...'
    js_ids = set(re.findall(r'document\.getElementById\([\'"]([a-zA-Z0-9_-]+)[\'"]\)', js_content))

    # Extract all id="..." or '...' in HTML
    html_ids = set(re.findall(r'id=[\'"]([a-zA-Z0-9_-]+)[\'"]', html_content))

    missing_in_html = js_ids - html_ids

    print("========================================")
    print("DOM INTEGRITY REPORT (PYTHON): d-lamp Mode")
    print("========================================")
    print(f"Total Unique IDs queried in JS: {len(js_ids)}")
    print(f"Total Unique IDs defined in HTML: {len(html_ids)}")
    print("----------------------------------------")

    if missing_in_html:
        print("[FAIL] CRITICAL FAILURE: JS references these IDs, but they do NOT exist in HTML:")
        for id_name in missing_in_html:
            print(f"  - {id_name}")
        sys.exit(1)
    else:
        print("[SUCCESS] All JS DOM element query IDs are present in the HTML DOM!")
        sys.exit(0)
except Exception as e:
    print(f"[FATAL] FATAL ERROR running verification: {str(e)}")
    sys.exit(1)
